aesara/link/numba/dispatch/basic.py

```diff
@@ -350,7 +352,63 @@ def numba_const_convert(data, dtype=None, **kwargs):
 
 def numba_funcify(obj, node=None, storage_map=None, **kwargs) -> Callable:
     """Convert `obj` to a Numba-JITable object."""
-    return _numba_funcify(obj, node=node, storage_map=storage_map, **kwargs)
+    numba_py_fn = None
+    if config.DISABLE_NUMBA_CACHE:
+        numba_py_fn = _numba_funcify(obj, node=node, storage_map=storage_map, **kwargs)
+    else:
+        node_key = make_node_key(node)
+
+        if node_key:
+            numba_py_fn = check_cache(node_key)
+        if node_key is None or numba_py_fn is None:
+            # We could only ever return the function source in our dispatch
+            # implementations. That way, we can compile directly to the on-disk
+            # modules only once.
+            numba_py_fn = _numba_funcify(
+                obj, node=node, storage_map=storage_map, **kwargs
+            )
+
+            # This will determine on-disk module name to be generated for
+            # `numba_py_src` and return the corresponding Python function
+            # object using steps similar to
+            # `aesara.link.utils.compile_function_src`.
+            if node_key:
+                numba_py_fn = add_to_cache(node_key, numba_py_fn)
+
+        # TODO: Presently numba_py_fn is already jitted.
+        # numba_fn = numba_njit(numba_py_fn)
+    return cast(Callable, numba_py_fn)
```

Comment:
FYI: I think the tests are failing because we haven't finished refactoring the rest of the code so that it's aware of `numba_funcify` now returning un-`njit`ed Python functions.  For example, the `test_config_options_cached` is expecting `numba_mul_fn` to be a Numba `CPUDispatcher` object and not a plain Python function object.